CopilotKit/packages/backend/src/utils/stream.ts

```diff
@@ -145,55 +149,61 @@ export function copilotkitStreamInterceptor(
             writeChatCompletionEnd(controller);
             await cleanup(controller);
             return;
-          } else if (debug) {
+          } // done == true (terminal case)
+
+          if (debug) {
             console.log("data: " + JSON.stringify(value) + "\n\n");
           }
 
-          let mode: "function" | "message" = value.choices[0].delta.tool_calls
-            ? "function"
-            : "message";
+          type Mode = { type: "function"; toolCall: ToolCallPayload } | { type: "message" };
 
-          const index = (value.choices[0].delta.tool_calls?.[0]?.index || 0) as number;
+          let mode: Mode;
+          const maybeToolCall = value.choices[0].delta.tool_calls?.[0];
+          if (maybeToolCall) {
+            mode = { type: "function", toolCall: maybeToolCall };
+          } else {
+            mode = { type: "message" };
+          }
 
-          // We are in the middle of a function call and got a non function call chunk
+          const nextChunkIndex = mode.type === "function" ? mode.toolCall.index : null;
+          // If We are in the middle of a function call and got a non function call chunk
           // or a different function call
           // => execute the function call first
-          if (executeThisFunctionCall && (mode != "function" || index != currentFnIndex)) {
+          if (
+            executeThisFunctionCall &&
+            (mode.type != "function" || nextChunkIndex != currentFnIndex)
+          ) {
             await flushFunctionCall(controller);
           }
-
-          currentFnIndex = index;
+          currentFnIndex = nextChunkIndex;
 
           // if we get a message, emit the content and continue;
-          if (mode === "message") {
+          if (mode.type === "message") {
             if (value.choices[0].delta.content) {
               writeChatCompletionChunk(controller, value);
             }
             continue;
           }
+
           // if we get a function call, emit it only if we don't execute it server side
-          else if (mode === "function") {
+          else if (mode.type === "function") {
             // Set the function name if present
-            if (value.choices[0].delta.tool_calls?.[0]?.function?.name) {
-              functionCallName = value.choices[0].delta.tool_calls![0].function.name!;
+            const maybeFunctionName = mode.toolCall.function.name;
+            if (maybeFunctionName) {
+              functionCallName = maybeFunctionName;
```

Comment:
here and below, we repeat ourselves less by defining an intermediate variable, checking if it's null/undefined, and then proceeding